cms/src/test/java/uk/nhs/digital/externalstorage/workflow/externalFileUnpublish/ExternalFileUnpublishTaskTest.java

```diff
@@ -42,22 +54,56 @@ public void setUp() throws RepositoryException {
         Repository repository = MockJcr.newRepository();
         session = repository.login();
         rootNode = session.getRootNode();
+        testFolder = rootNode.addNode("lorem-ipsum", NT_FOLDER);
         externalFileUnpublishTask = new ExternalFileUnpublishTask();
-        externalFileUnpublishTask.setS3Connector(s3Connector);
 
         given(workflowContext.getInternalWorkflowSession()).willReturn(session);
         externalFileUnpublishTask.setWorkflowContext(workflowContext);
     }
 
     @Test
-    public void shouldCallS3() throws WorkflowException, RepositoryException {
-        Node docNode = rootNode.addNode("test-document");
-        addVariantNode(docNode, PUBLISHED);
-        MockJcr.setQueryResult(session, getQueryResults());
+    public void shouldUnpublishResources() throws WorkflowException, RepositoryException {
+        MockJcr.addQueryResultHandler(session, searchResourcesHandle(testExternalReference));
+
+        executeTask(getDocumentHandleWithVariant(getRandomDocumentType(), PUBLISHED), externalFileUnpublishTask);
+
+        then(s3Connector).should().unpublishResource(testExternalReference);
+        then(s3Connector).should(never()).publishResource(anyString());
+    }
+
+    @Test
+    public void shouldUnpublishPublicationAndItsDatasetsResources() throws WorkflowException, RepositoryException {
+        // add Dataset to JCR
+        Node dataset = getDataset();
+        Node publication = getPublication();
 
-        executeTask(docNode, externalFileUnpublishTask);
+        MockJcr.addQueryResultHandler(session, searchResourcesHandle(testExternalReference));
+        MockJcr.addQueryResultHandler(session, searchDatasetsHandle(dataset));
+
+        // when unpublishing upcoming publication
+        executeTask(getDocumentHandleWithVariant(publication, PUBLISHED), externalFileUnpublishTask);
+
+        // should be executed once for publication resource and once for dataset resource
+        then(s3Connector).should(times(2)).unpublishResource(testExternalReference);
```

Comment:
I think this should use different `testExternalReference` values for dataset and publication. As it is, the test doesn't prevent a bug where the same reference is used for each of the external resources that need publishing.

Reply:
adding logic to return different search results for different node queries